server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java

```diff
@@ -90,10 +91,16 @@ boolean recoverFromStore(final IndexShard indexShard) {
             RecoverySource.Type recoveryType = indexShard.recoveryState().getRecoverySource().getType();
             assert recoveryType == RecoverySource.Type.EMPTY_STORE || recoveryType == RecoverySource.Type.EXISTING_STORE :
                 "expected store recovery type but was: " + recoveryType;
-            return executeRecovery(indexShard, () -> {
+            final PlainActionFuture<Boolean> future = PlainActionFuture.newFuture();
+            final ActionListener<Boolean> recoveryListener = recoveryListener(indexShard, future);
+            try {
                 logger.debug("starting recovery from store ...");
                 internalRecoverFromStore(indexShard);
-            });
+                recoveryListener.onResponse(true);
+            } catch (Exception e) {
+                recoveryListener.onFailure(e);
+            }
+            return future.actionGet();
```

Comment:
It's a bit of a dirty pattern to use the `PlainActionFuture` here but I didn't want to make the other recoveries fully async and blow-up the scope of this PR.
In practice the future resolves in-sync here anyway and we never truely block on this line.

Reply:
> executeRecovery()

I tried, I really did. But the problem is that there's not more synchronous `Runnable` to execute as the callback. So the only thing you can do to share code between the sync. and asyn paths taken by other restores and the repo is to pass a `Consumer<ActionListener<Boolean>>` which turned out to really not add anything in readability and just makes everything more complex.
I think here the next step should just be to make these things async as well in terms of code since it all just executes on a single thread anyway. Same as with the CCR change I just didn't want to change the flow for this PR.

Reply:
Ok, thanks for explaining.